### PR TITLE
[SIL] Make @_silgen_name and @_cdecl functions immune to some optimizations

### PR DESCRIPTION
**Pull request by gparker42**
@_silgen_name and @_cdecl functions are assumed to be referenced from
C code. Public and internal functions marked as such must not be deleted
by the optimizer, and their C symbols must be public or hidden respectively.

rdar://33924873, SR-6209

### CODE REVIEW

**Comment by gparker42** [comment]
@swift-ci please test

**Review by eeckstein — APPROVED** [review]
lgtm!

**Comment by gparker42** [comment]
@swift-ci please smoke test

**Comment by gparker42** [comment]
@swift-ci please smoke test

**Comment by gparker42** [comment]
@swift-ci please smoke test

File: test/IRGen/asmname.swift [hunk 1]
```diff
@@ -1,9 +1,50 @@
-// RUN: %target-swift-frontend -assume-parsing-unqualified-ownership-sil %s -emit-ir | %FileCheck %s
+// RUN: %target-swift-frontend -assume-parsing-unqualified-ownership-sil %s -emit-ir -whole-module-optimization | %FileCheck %s
```
Comment (jrose-apple) [inline test/IRGen/asmname.swift]: `-whole-module-optimization` isn't a frontend option.

Comment (gparker42) [reply test/IRGen/asmname.swift]: Unless I made a mistake I definitely saw -whole-module-optimization changing the output of this test, and I do see other tests that use this combination. I'll look again.

Comment (gparker42) [reply test/IRGen/asmname.swift]: Nope, I must have made a mistake somewhere. I see no output difference now.

Comment (gparker42) [reply test/IRGen/asmname.swift]: https://github.com/apple/swift/pull/12714

**Pull request merged**